[ios] Improve TTS voice selection

fabwu opened the pull request:
This commit allows a user to select more voices (e.g. English (United States), English (India)). Currently, it's only possible to select a subset of the available voices on iOS. For example, if a user selects English as TTS language, an Australian voice is selected because en_AU comes before en_US in the voice list of iOS.

Fixes #6840
Fixes #3222
Fixes #2178
Fixes #7604

kirylkaveryn commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech.mm:
```diff
@@ -286,16 +288,12 @@ + (void)removeObserver:(id<MWMTextToSpeechObserver>)observer {
-    return "";
+  NSString *nsLocaleString = [NSString stringWithUTF8String: localeString.c_str()];
+  NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier: nsLocaleString];
+  NSString *localizedName = [locale localizedStringForLocaleIdentifier:nsLocaleString];
```
```suggestion
  NSString * localizedName = [locale localizedStringForLocaleIdentifier:nsLocaleString];
```

fabwu replied:
Mmh I never get used to Objective-C syntax ^^ but I stick to the convention.

biodranik commented on iphone/Maps/UI/Settings/MWMTTSSettingsViewController.mm:
```diff
@@ -242,8 +241,7 @@ - (void)viewDidLoad
         !savedLanguage.empty())
-      m_languages.emplace_back(
-          make_pair(savedLanguage, translatedTwine(bcp47ToTwineLanguage(nsSavedLanguage))));
+      m_languages.emplace_back(make_pair(savedLanguage, translateLocale(savedLanguage)));
```
```suggestion
      m_languages.emplace_back(savedLanguage, translateLocale(savedLanguage));
```

fabwu replied:
I really have to look into modern C++. Thanks!

biodranik commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech.mm:
```diff
@@ -286,16 +288,12 @@ + (void)removeObserver:(id<MWMTextToSpeechObserver>)observer {
+  NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier: nsLocaleString];
+  NSString *localizedName = [locale localizedStringForLocaleIdentifier:nsLocaleString];
+  localizedName = [localizedName capitalizedString];
+  return std::string([localizedName UTF8String]);
```
ObjC getters without parameters are easier to read as properties.
```suggestion
  return std::string(localizedName.UTF8String);
```

fabwu replied:
Thanks!

biodranik commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech.mm:
```diff
@@ -33,8 +33,10 @@
         // Twine names are equal. Make a pair: bcp47 name, localized name.
-        result.emplace_back(make_pair(lang.second, p.second));
-        break;
+        auto pair = make_pair(lang.second, tts::translateLocale(lang.second));
+        if(std::find(result.begin(), result.end(), pair) == result.end()) {
+          result.emplace_back(pair);
+        }
```
Change above:
```
for (auto const & p : kLanguageList)
// to
for (auto const & [twineCode, languageName] : kLanguageList)
```

```suggestion
        auto pair = std::make_pair(lang.second, tts::translateLocale(lang.second));
        if (std::find(result.begin(), result.end(), pair) == result.end())
          result.emplace_back(std::move(pair));
```

fabwu replied:
Thanks didn't know the `std::move`

biodranik reviewed (commented):
Cool, thanks!

fabwu commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech+CPP.h:
```diff
@@ -14,5 +14,5 @@
 namespace tts
 {
-std::string translatedTwine(std::string const & twine);
+std::string translateLocale(std::string const & twine);
```
Thanks good catch!

fabwu commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech.mm:
```diff
@@ -286,16 +288,12 @@ + (void)removeObserver:(id<MWMTextToSpeechObserver>)observer {
-    return "";
+  NSString *nsLocaleString = [NSString stringWithUTF8String: localeString.c_str()];
+  NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier: nsLocaleString];
+  NSString *localizedName = [locale localizedStringForLocaleIdentifier:nsLocaleString];
```
I'd love to write a test but is there a way to call this method from Swift oder do you suggest a test in Objective-C?

kirylkaveryn replied:
Swift will not be working with the cpp (until we upgrade to the latest xcode and up the target to ios 15 because there will no way back to the xcode 14 - https://www.swift.org/documentation/cxx-interop/#enabling-c-interoperability).
So the only way to write tests for this method is objc.

fabwu replied:
Ok I wrote some tests in Objective-C. It was easier than expected...

I also moved the test classes to the same group as their respective classes in the project. The Tests group basically mimics the project folder structure:

<img width="350" alt="image" src="https://github.com/organicmaps/organicmaps/assets/6050639/86dc986e-f7b8-4fdf-af6b-f39001a66254">

I do this in my large projects to organize my test files but I don't know if you like it?

fabwu commented:
I've changed two more things:

1. I removed the hard-coded Alex voice for `en-US` because I wasn't able to select another voice if the Alex voice is installed.

2. The first entry of the voice list was `English (Australian)` (see screenshot). I changed it to `English (United States)`. Now the order is `English (United States)`, default voice language depending on the iOS settings and then the selected language.

<image width="40%" src="https://github.com/organicmaps/organicmaps/assets/6050639/2ca98106-4eef-443e-b020-6e1ddd9faeac">

kirylkaveryn commented on iphone/Maps/Maps.xcodeproj/project.pbxproj:
```diff
@@ -3759,6 +3794,9 @@
+					ED097E752BB80C320006ED01 = {
+						LastSwiftMigration = 1530;
+					};
 				};
```
I think this autogenerated code can be removed.

fabwu replied:
Ah sorry I don't know how this got generated. I removed all lines.

kirylkaveryn commented on iphone/Maps/Maps.xcodeproj/project.pbxproj:
```diff
@@ -4689,12 +4728,15 @@
 			baseConfigurationReference = 347526FA1DC0B00F00918CF5 /* common-debug.xcconfig */;
 			buildSettings = {
 				BUNDLE_LOADER = "$(TEST_HOST)";
+				CLANG_ENABLE_MODULES = YES;
```
Please try to remove all this autogenerated code from the `buildSettings` and run tests.
It should be OK.

biodranik replied:
+1, this is not needed. We don't use modules.

kirylkaveryn commented on iphone/Maps/Maps.xcodeproj/project.pbxproj:
```diff
@@ -4689,12 +4728,15 @@
 				PRODUCT_BUNDLE_IDENTIFIER = app.organicmaps.debug.tests;
 				PROVISIONING_PROFILE_SPECIFIER = "";
+				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
+				SWIFT_VERSION = 5.0;
```
Same

kirylkaveryn commented on iphone/Maps/Maps.xcodeproj/project.pbxproj:
```diff
@@ -4704,11 +4746,13 @@
 				"DEVELOPMENT_TEAM[sdk=macosx*]" = 9Z6432XD7L;
 				PRODUCT_BUNDLE_IDENTIFIER = app.organicmaps.release.tests;
 				PROVISIONING_PROFILE_SPECIFIER = "";
+				SWIFT_VERSION = 5.0;
```
Here too

biodranik commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech.mm:
```diff
@@ -26,15 +26,15 @@
-        result.emplace_back(make_pair(lang.second, p.second));
-        break;
+        auto pair = make_pair(bcp47Voice, tts::translateLocale(bcp47Voice));
+        if(std::find(result.begin(), result.end(), pair) == result.end())
```
```suggestion
        if (std::find(result.begin(), result.end(), pair) == result.end())
```

fabwu replied:
Ah sorry I missed that

biodranik commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech.mm:
```diff
@@ -118,6 +118,9 @@ - (void)dealloc {
 }
 - (std::vector<std::pair<std::string, std::string>>)availableLanguages { return _availableLanguages; }
+- (std::pair<std::string, std::string>)standardLanguage {
+  return make_pair(kDefaultLanguage.UTF8String, tts::translateLocale(kDefaultLanguage.UTF8String));
```
```suggestion
  return std::make_pair(kDefaultLanguage.UTF8String, tts::translateLocale(kDefaultLanguage.UTF8String));
```

fabwu replied:
done

biodranik commented on iphone/Maps/Core/TextToSpeech/MWMTextToSpeech.mm:
```diff
@@ -26,15 +26,15 @@
-        // Twine names are equal. Make a pair: bcp47 name, localized name.
-        result.emplace_back(make_pair(lang.second, p.second));
-        break;
+        auto pair = make_pair(bcp47Voice, tts::translateLocale(bcp47Voice));
```
```suggestion
        auto pair = std::make_pair(bcp47Voice, tts::translateLocale(bcp47Voice));
```

fabwu replied:
done

biodranik commented on iphone/Maps/Maps.xcodeproj/project.pbxproj:
```diff
@@ -4704,11 +4746,13 @@
 			baseConfigurationReference = 347526FB1DC0B00F00918CF5 /* common-release.xcconfig */;
 			buildSettings = {
 				BUNDLE_LOADER = "$(TEST_HOST)";
+				CLANG_ENABLE_MODULES = YES;
```
It should be easy to remove some lines when amending the commit using Fork, Github App or other Git app with UI.

biodranik reviewed (commented):
Good, thanks!

biodranik reviewed (commented):
A bit of final minor nitpicking )

fabwu commented on iphone/Maps/UI/Settings/MWMTTSSettingsViewController.mm:
```diff
@@ -219,16 +219,15 @@ - (void)viewDidLoad
   auto const & v = tts.availableLanguages;
   NSAssert(!v.empty(), @"Vector can't be empty!");
-  pair<string, string> const standart = v.front();
+  pair<string, string> const standart = tts.standardLanguage;
   m_languages.push_back(standart);
```
That doesn't work because the variable is used below. I changed it to `standard` though.

fabwu commented on iphone/Maps/Tests/Core/TextToSpeech/MWMTextToSpeechTests.mm:
```diff
@@ -0,0 +1,28 @@
+- (void)testAvailableLanguages {
+  MWMTextToSpeech * tts = [MWMTextToSpeech tts];
+  std::vector<std::pair<std::string, std::string>> langs = tts.availableLanguages;
+  auto defaultLang = std::make_pair("en-US", "English (United States)");
```
Ok I fixed everything as suggested :)

biodranik reviewed (approved):
LGTM, thanks!

@kirylkaveryn @vng any comments?
